cirq-superstaq/cirq_superstaq/job.py

```diff
@@ -149,42 +223,137 @@ def repetitions(self) -> int:
             self._refresh_job()
 
-        return self._job["shots"]
+        return self._job[first_job_id]["shots"]
+
+    @overload
+    def _get_circuits(self, circuit_type: str, index: int) -> cirq.Circuit:
+        ...
+
+    @overload
+    def _get_circuits(
+        self, circuit_type: str, index: None = None
+    ) -> Union[
+        cirq.Circuit, List[cirq.Circuit]
+    ]:  # Change return to `List[cirq.Circuit]` after deprecation
+        ...
 
-    def compiled_circuit(self) -> cirq.Circuit:
-        """Gets the compiled circuit that was submitted for this job.
+    def _get_circuits(
+        self, circuit_type: str, index: Optional[int] = None
+    ) -> Union[cirq.Circuit, List[cirq.Circuit]]:
+        """Retrieves the corresponding circuits to `circuit_type`.
+
+        Args:
+            circuit_type: The kind of circuits to retrieve. Either "input_circuit" or
+                "compiled_circuit".
+            index: The index of the circuit to retrieve.
 
         Returns:
-            The compiled circuit.
+            A single circuit or list of circuits.
         """
-        if "compiled_circuit" not in self._job:
+        if circuit_type not in ("input_circuit", "compiled_circuit"):
+            raise ValueError("The circuit type requested is invalid.")
+
+        job_ids = self._job_id.split(",")
+        if not all(
+            job_id in self._job and self._job[job_id].get(circuit_type) for job_id in job_ids
+        ):
             self._refresh_job()
 
-        return css.deserialize_circuits(self._job["compiled_circuit"])[0]
+        if index is None:
+            serialized_circuits = [self._job[job_id][circuit_type] for job_id in job_ids]
+            return [css.deserialize_circuits(serialized)[0] for serialized in serialized_circuits]
+
+        serialized_circuit = self._job[job_ids[index]][circuit_type]
+        return css.deserialize_circuits(serialized_circuit)[0]
+
+    @overload
+    def compiled_circuits(self, index: int) -> cirq.Circuit:
+        ...
+
+    @overload
+    def compiled_circuits(
+        self, index: None = None
+    ) -> Union[
+        cirq.Circuit, List[cirq.Circuit]
+    ]:  # Change return to `List[cirq.Circuit]` after deprecation
+        ...
 
-    def input_circuit(self) -> cirq.Circuit:
-        """Gets the original circuit that was submitted for this job.
+    def compiled_circuits(
+        self, index: Optional[int] = None
+    ) -> Union[cirq.Circuit, List[cirq.Circuit]]:
+        """Gets the compiled circuits that were submitted for this job.
+
+        Args:
+            index: The index of the circuit to retrieve.
 
         Returns:
-            The submitted input circuit.
+            A single compiled circuit or list of compiled circuits.
         """
-        if "input_circuit" not in self._job:
-            self._refresh_job()
+        if index is not None:
+            gss.validation.validate_integer_param(index, min_val=0)
```

Comment:
can we move this to `_get_circuits()`?

Reply:
Yes, I actually have it like that in the `qss` implementation